test-data/unit/check-classes.test

```diff
@@ -2065,3 +2065,167 @@ class A: pass
 class B(object, A): # E: Cannot determine consistent method resolution order (MRO) for "B"
     def readlines(self): pass
     __iter__ = readlines
+
+[case testSelfTypeInstance]
+from typing import TypeVar
+
+T = TypeVar('T')
+
+class A:
+    def copy(self: T) -> T: ...
+
+class B(A):
+    pass
+
+reveal_type(A().copy)  # E: Revealed type is 'def () -> __main__.A'
+reveal_type(B().copy)  # E: Revealed type is 'def () -> __main__.B'
+reveal_type(A().copy())  # E: Revealed type is '__main__.A'
+reveal_type(B().copy())  # E: Revealed type is '__main__.B'
+# FIX: Why partial type?
+# reveal_type(A.copy(A()))
+# reveal_type(B.copy(B()))
+
+
+[case testSelfTypeReturn]
+from typing import TypeVar, Type
+
+T = TypeVar('T')
+
+def _type(self: T) -> Type[T]: ...
+
+class A:
+    def copy(self: T) -> T:
+        if B():
+            return A()
```

Comment:
Try to use the `# E: error message` convention for the error messages for readability and consistency. If it doesn't work because of all the notes (it shouldn't be a problem here), use a `# Error` comment on lines that generate an error to make it easier to match errors to lines. You can leave the `main: note: In ...` in the `[out]` section and add `# E: ...` comments here.